### PR TITLE
Added limitations section first draft

### PR DESCRIPTION
**Pull request by andykirk**
As discussed in #53

+@ausi

***
<a href="https://pr-preview.s3.amazonaws.com/andykirk/cq-usecases/pull/64.html" title="Last updated on Mar 2, 2018, 8:38 AM GMT (2b6c656)">Preview</a> | <a href="https://pr-preview.s3.amazonaws.com/WICG/cq-usecases/64/4077161...andykirk:2b6c656.html" title="Last updated on Mar 2, 2018, 8:38 AM GMT (2b6c656)">Diff</a>

### CODE REVIEW

File: index.bs
Comment (ausi) [inline index.bs]: I think we should add information about the additional reflows here too.
Somethink like: *JavaScript solutions also create additional network traffic, require additional parsing and perform an additional reflow after every layout that affects the matching state of a container query.*

@andykirk do you want us to discuss this right here?
Or should we merge your pull request first and I create a sepearte pull request for this afterwards?

Comment (andykirk) [reply index.bs]: Hi. Not sure / don't mind. Happy to discuss here as @beep suggested, but also happy with your suggested change. Maybe we can amend this PR?

**Comment by ausi** [comment]
> Without the use of JavaScript, iframes require that the contents of the iframe are stored in a <em>separate HTML document</em> rather than being part of the current document.

Since we have the `srcdoc` attribute for iframes nowadays (except in IE/Edge), this point seems to no longer be 100% correct.

Should we add that iframes are also problematic because they stop the styles from being inherited, and thus every component has to set font family/size/color again?

File: index.bs
Comment (ausi) [reply index.bs]: Amending seems right to me.
Can you commit my change and push it to your branch?
(As I have no push rights to your repo)

Comment (andykirk) [reply index.bs]: Sure,

**Comment by andykirk** [comment]
@ausi  Good point. I hadn't considered that. Maybe re-phrase / clarify it for _"some browsers"_?

Without the use of JavaScript, <del>iframes</del><ins>some browsers</ins> require that the contents of the iframe are stored in a separate HTML document rather than being part of the current document.

?

File: index.bs
Comment (andykirk) [reply index.bs]: @ausi PR amened. This part of the conversation now 'outdated'.

**Comment by ausi** [comment]
:+1:

**Comment by andykirk** [comment]
Amended again.

Re. style inheritance, what about:

3. The contents of an iframe do not inherit any styles from their parent document, forcing the styles used by the 'module' to be redeclared.

File: index.bs
Comment (tomhodgins) [inline index.bs]: Despite having some security implications, iframes on mobile safari on iOS _are_ affected by the document contained inside them. By default, the height of the iframe expands to match the height of the embedded document.

Comment (tomhodgins) [reply index.bs]: ^ this was referring to this wording:

> The contents of an iframe do not natively affect the dimensions of the iframe itself.

Comment (andykirk) [reply index.bs]: Ok, thanks. Will consider how to re-phrase that.

Comment (andykirk) [reply index.bs]: @tomhodgins - I've re-phrased. This part of the conversation is now 'outdated'.

**Review by tomhodgins — APPROVED** [review]
I'm okay with the wording here since we can always edit this after pulling it in too!

My 2¢ is that “The only known exception to this at the time of writing is iOS Safari” being at the end of its paragraph makes it sound to me like iOS safari is the exception to this part: “The contents of a container that is the subject of a Container Query are still expected to be able to affect the layout of its parents in the <em>opposite</em> dimension of the query. …” except in iOS Safari?

Plus, relating to that limitation part: have we discussed and agreed on the limitation that it _shouldn't_ be able to affect the original queried element in the same sense of the query? I saw one suggestion somewhere related to that, but I'm not sure that idea has been scrutinized by the group yet.

**Comment by andykirk** [comment]
@tomhodgins you're right - I was being hasty. I can move that to where it makes more sense.

I'm not sure about your 2nd point. I thought that if the contents of the 'module' do not / cannot affect the size of the container itself, we're making the whole thing a bit useless? E.g. if layout changes in response to container width - the height of the container _must_ be able to adapt to the new content height or you'd have gaps / overflow issues?

That was my understanding.

**Comment by beep** [comment]
> Plus, relating to that limitation part: have we discussed and agreed on the limitation that it shouldn't be able to affect the original queried element in the same sense of the query? I saw one suggestion somewhere related to that, but I'm not sure that idea has been scrutinized by the group yet.

> I thought that if the contents of the 'module' do not / cannot affect the size of the container itself, we're making the whole thing a bit useless? E.g. if layout changes in response to container width - the height of the container must be able to adapt to the new content height or you'd have gaps / overflow issues?

I might suggest pulling this point out into an issue, if it merits discussion _and_ it doesn’t directly affect/block this PR.

**Comment by andykirk** [comment]
Re-phrased content sizing exception.

**Comment by andykirk** [comment]
Comments for reference:

https://github.com/WICG/cq-usecases/issues/49#issuecomment-360551606
https://github.com/WICG/cq-usecases/issues/49#issuecomment-360557485

Happy for this to be a separate issue. I don't think it blocks this PR? AS @tomhodgins say's, we can edit this post-PR.

File: index.bs
Comment (ausi) [inline index.bs]: The “two significant drawbacks” are now three 🤓

**Comment by beep** [comment]
@tomhodgins @hereinthehive, just checking: is this ready to merge? If not, are there specific items that need to be addressed by @ausi and/or @andykirk? Thanks!

File: index.bs
Comment (andykirk) [reply index.bs]: Doh! Of course, good spot.
Fixed.

**Review by ausi — APPROVED** [review]

**Comment by tomhodgins** [comment]
I'm good with this PR, as long as we create that issue to figure out what kind of containment we might want to put in here. Thanks for your writing and edits @andykirk 🙌

**Comment by beep** [comment]
> I'm good with this PR

Marvelous, thanks. Then merge away, sir!

> as long as we create that issue to figure out what kind of containment we might want to put in here.

Sounds grand. Would you mind writing up that issue, @tomhodgins?

File: index.bs
Comment (marcoscaceres) [inline index.bs]: You should probably briefly say what the fab four technique is. like:

"such as the fab four technique, which bla bla briefly bla bla" 

It's unlikely the reader has heard of this (I'd never heard of it)... but the reader should be able to get the basic gist of what it is, and then continue reading... with the *option* of clicking the link if they want more info.

File: index.bs
Comment (marcoscaceres) [inline index.bs]: Nit: this is screaming to be a bunch of list items, a definition list (with a brief name for each workaround), or even small sub sections. 

 * complex media queries
 * Flex, Grid, and ???: 
 * iframes

File: index.bs
Comment (marcoscaceres) [inline index.bs]: nit: `code` instead of `strong`?

File: index.bs
Comment (marcoscaceres) [reply index.bs]: We want to highlight how ridiculously complex a media query could be, so showing one here would help.

File: index.bs
Comment (marcoscaceres) [inline index.bs]: It also incurs the additional overhead of unnecessarily spinning up an additional nested browsing context, which can degrade performance (lots of iframes would be bad!).

File: index.bs
Comment (marcoscaceres) [inline index.bs]: "natively affect" is unclear. Its also not clear what we mean by Safari behaving differently.

File: index.bs
Comment (marcoscaceres) [inline index.bs]: I guess it also builds up a whole new CSSOM in each iframe, which would be expensive too. 

Has anyone been crazy enough to use this technique in practice? If so, we should list sites if we know any.

File: index.bs
Comment (marcoscaceres) [inline index.bs]: Please break this up into two sentences.

**Review by marcoscaceres — CHANGES_REQUESTED** [review]
Drive by comments 🚙💨.

**Comment by marcoscaceres** [comment]
Generally speaking, this is a good start, but I'd like to see it made a bit stronger before you merge it.  

I suggest strengthening this a little bit more and pinging @tantek to review it. If @tantek is like, "yeah, that seems seriously bad"... then we know we are onto something. 

@tantek might also have suggestion for how to make sure the arguments are well targeted at CSS/Browser folks.

File: index.bs
Comment (marcoscaceres) [reply index.bs]: Oh, with the javascript stuff... we should say that there are like 20 JS solutions out there - and link to the list. If you know that a solution is being used on a really popular site (i.e., one with a few million users), please say which one.

File: index.bs
Comment (andykirk) [reply index.bs]: Ok, I’ll give it a go. I want to avoid bloat, but you’re right, it needs context. I guess I assumed the link to the technique was enough.

File: index.bs
Comment (andykirk) [reply index.bs]: First point - that’s fine, it’s just how I write. I can change that.

Second point - I thought the original work in the Use Cases section inferred that complexity.
The [article by Mat Marquis](https://alistapart.com/article/container-queries-once-more-unto-the-breach) does illustrate it further. Perhaps a link to that? If not, perhaps some else can find one in the wild or write one? I don’t really have the capability to to come up with a good example at the moment and I don’t want to hold things up.
Again, I’d like to avoid bloating this if possible.

File: index.bs
Comment (andykirk) [reply index.bs]: I used `stong` as it’s like and inline heading, like the points above. I’ll swap it out as part of the change to a `dl`.

File: index.bs
Comment (andykirk) [reply index.bs]: Ok. Someone want to come up with some wording for that for me to include?

File: index.bs
Comment (andykirk) [reply index.bs]: Ok, I can expand on that. @tomhodgins do you have any resources or evidence for the Safari iOS thing?

File: index.bs
Comment (andykirk) [reply index.bs]: I don’t know. Anyone?

File: index.bs
Comment (andykirk) [reply index.bs]: First point - fair enough.

Second point - can we include a list on the Wiki so I can link to it? Anyone care to make that list? @tomhodgins?

**Comment by andykirk** [comment]
Thanks @marcoscaceres, I’ve responded to each of your points and I’m happy to rework it.
It will cause a delay though. If it were merged others could then try to tackle some of the points as separate PRs.

Cheers

**Comment by marcoscaceres** [comment]
> If it were merged others could then try to tackle some of the points as separate PRs.

Sure, up to you all how you want to work. Two options: keep iterating and using the Previews at the top of this pull request to share and review the changes, or merge early/often and iterate. 

I'm personally a fan of the former (keep hacking on a single PR), but the other way works fine too.

**Review by marcoscaceres — APPROVED** [review]

**Comment by marcoscaceres** [comment]
(approving changes, but please fix the small things if possible :) )

**Comment by andykirk** [comment]
Thanks @marcoscaceres - will fix as much as I can. Some of the other things will require more work/discussion.

File: index.bs
Comment (eeeps) [reply index.bs]: @wilto maybe https://twitter.com/wilto/status/728211240384929793

File: index.bs
Comment (eeeps) [reply index.bs]: Here's the list of JS solutions on the wiki https://github.com/WICG/cq-usecases/wiki/Speculative-Prolyfills

As for usage, the publicwww links that @tomhodgins showed me a while back are the best window into usage that I've yet found https://github.com/WICG/cq-usecases/issues/41.

The most popular site on the three lists there (for EQCSS, cq-prolyfill, and css-element-queries) is harvard.edu, with (Alexa?) rank #702.

File: index.bs
Comment (eeeps) [reply index.bs]: There isn't yet a standards-based way to declaratively tell an iframe to auto-size to fit its contents (see https://github.com/w3c/csswg-drafts/issues/1771). Can we leave it at that, rather than getting into implementation-specific hacks?

File: index.bs
Comment (eeeps) [reply index.bs]: I think @marcoscaceres already nailed it. How about just,

> This necessitates a new document composition paradigm that's very different from the methods currently employed by developers, and incurs the additional overhead of generating an additional nested browsing context, which degrades performance.

File: index.bs
Comment (andykirk) [reply index.bs]: That would be my preference. @tomhodgins ? @eeeps was the intention to include the link to the draft spec you cited?

File: index.bs
Comment (eeeps) [reply index.bs]: Oops! I see you already basically did this in https://github.com/WICG/cq-usecases/pull/64/commits/a009cf91dd5b565449df3f74302473c3b771abb6 – apologies.

File: index.bs
Comment (eeeps) [inline index.bs]: See my comment on the old commit here https://github.com/WICG/cq-usecases/pull/64/files/28f7e75daa6f74948281a2b4a7ad26cb578cfe35#r168499068 – I don't know if it's worth mentioning Safari's non-standard behavior.

File: index.bs
Comment (tomhodgins) [reply index.bs]: It's been happening on iOs as long as I can remember, I looked it up the first time I saw it and found articles, but lately I can't find anything on Google - so I wrote my own test:

```html
<meta name=viewport content="width=device-width, initial-scale=1">
<iframe></iframe>
<iframe src="http://staticresource.com/"></iframe>
<iframe src="http://eqcss.com/"></iframe>
```

Screenshots: https://imgur.com/a/SKjp4

File: index.bs
Comment (andykirk) [reply index.bs]: Thanks @eeeps. I realised after I commented that I could just use that. See [
a009cf9](https://github.com/WICG/cq-usecases/pull/64/commits/a009cf91dd5b565449df3f74302473c3b771abb6)

File: index.bs
Comment (tomhodgins) [reply index.bs]: Might want to ask [Hugo Giraudel](https://twitter.com/HugoGiraudel) -> https://github.com/edenspiekermann/iframify

I believe they this solution on style guides

File: index.bs
Comment (tomhodgins) [reply index.bs]: There's also a dozen or so listed on http://whoisworkingoncontainerqueries.com

File: index.bs
Comment (eeeps) [reply index.bs]: Ok so maybe:

> There currently is no standard, declarative way for an `iframe` to auto-size to fit its contents.
>
> Note: Mobile Safari auto-sizes `iframe`s by default, and there is [active discussion](https://github.com/w3c/csswg-drafts/issues/1771) within the CSSWG about whether or not `iframe` auto-sizing functionality should be standardized.
>
> The contents of a container that is the subject of a Container Query...

(the "Note:" at the beginning [does special Bikeshed things](https://tabatkins.github.io/bikeshed/#notes-etc) which create a green box to offset the note from the main text)

Comment (eeeps) [reply index.bs]: That is, only if, @tomhodgins, you think Mobile Safari's unique behavior and/or potential future standardized features are important to note. I favor keeping the text clean and the argument – iframes weren't built for this and aren't suited for this – simple/strong, but can see why you would want to include the fact that using them is sometimes possible in limited contexts.

Comment (tomhodgins) [reply index.bs]: I think it's worth mentioning that (in addition to other drawbacks of iframes) due to cross-browser inconsistencies with iframe sizing, using iframes isn't a good solution for embedding self-responsive layouts in other documents. To use it reliably you'd have to assign a height to the document or iframe, and in doing so you limit the kinds of layouts you can embed.

I don't know that saying which browsers act differently is important to that point, so we could probably leave browser names out, but if we can dig up a link where they can read more about it that would be helpful for those who want to learn more about why that's true.

Comment (beep) [reply index.bs]: I’d opt for more generic language here, personally. If only because it’s possible that if a specific (non-standard) implementation might change its behavior in the future.

Comment (andykirk) [reply index.bs]: @eeeps @tomhodgins @beep - sorry, dropped out for a while.
To sum, I’m going to change this:

> In the majority of browsers ... iOS Safari.

To this:

> There currently is no standard, declarative way for an iframe to auto-size to fit its contents.

And leave it at that?

Comment (beep) [reply index.bs]: @andykirk That sounds great to me.

Comment (eeeps) [reply index.bs]: @andykirk Me too.

Comment (andykirk) [reply index.bs]: Ok, I updated this.
I also re-worked the example/requirements but. I know we still need to discuss that, but I think it reads better now. Happy to change back or re-work if needed.

**Comment by andykirk** [comment]
Hi, sorry for the radio silence - I was struck down with the flu.

Is there anything here waiting on me?
The only thing I can think of was where @beep [suggests talking about the width - height thing in a seperate issue](https://github.com/WICG/cq-usecases/pull/64#issuecomment-365643834).

To be honest, it's very difficult describing the limitations there are with existing techniques when there isn't a solid set of requirements to compare them against. Sort of chicken / egg thing.

**Comment by beep** [comment]
@andykirk, thanks for the nudge. A few tiny copyedits from me:

- “`min-width` adn `max-width` to affect” s/b “`min-width` and `max-width` to affect”
- “based on the width of it’s parent” s/b “based on the width of its parent”
- “a child elements' `display` property” s/b “a child element’s `display` property”
- “based on a containers _width_” s/b “based on a container’s _width_”
- “incur additional overheads” s/b “incur additional overhead”

That’s all I’ve got. None of these feel like blockers to me…so unless anyone objects, I think this is ready to merge?

@andykirk, thanks so much to you (and everyone else participating here) for all your hard work getting this first draft in place. It’s looking really stellar to me!

**Comment by andykirk** [comment]
Thanks @beep. Apostrophes - ugh! Good spots. 

I can fix these bits when I get into the office tomorrow. 8am GMT. Too fried to do it now.

**Comment by beep** [comment]
@andykirk Hey, that sounds perfect. Thank you again!

**Comment by andykirk** [comment]
Done.

**Comment by beep** [comment]
@andykirk Looks grand to me. Thank you!

I’ll leave this open for a couple hours in case anyone has any final comments, but I think this is looking good to merge.

Thanks again for all your hard work, Andy. This feels like an _excellent_ first draft of the section.

**Review by beep — APPROVED** [review]

**Comment by tomhodgins** [comment]
Yes, thank you @andykirk for being patient with all these changes :D <3

**Review by eeeps — APPROVED** [review]

**Comment by beep** [comment]
🎉

**Comment by andykirk** [comment]
Thanks everyone 👍

**Comment by beep** [comment]
Thank _you_, @andykirk! And to everyone else who chimed in on this PR; I’m really excited about this.

**Comment by marcoscaceres** [comment]
Just had a read also. Looks great and reads nicely! Thanks everyone.

**Pull request merged**
